Kpi charts

stanleypliu opened the pull request:
**Ticket**

https://unep-wcmc.codebasehq.com/projects/labs/tickets/5

To set this up:
- Migrate
- `bundle install`
- Run `yarn install` 

`http://0.0.0.0:3000/kpis` should initially show a loading spinner because you don't have a Kpi instance in your db. To import the data, run `Kpi.instance` in your console, then once that's done refresh the page and the charts should show. 

Haven't covered the GA link yet, should I add that in?

WKocur reviewed (commented):
Looks top! Few comments:

- [ ] I've added a documentation link to one site, but the graph still shows none have documentation (I added this to species+)
- [ ] Can we use type=number for the number inputs? I don't know if number validation is on the backend, but means we can safely do sums etc. on this data
- [ ] Kpi -> KPI
- [ ] Can the KPI field be 'Key Performance Indicators (KPIs)'
- [ ] Rename KPI measurement to 'KPIs measurement (how and where)'
- [ ] Change % project led fields to select with 3 options (Informatics lead, co-design, in-house agency'

stanleypliu commented on app/controllers/projects_controller.rb:
```diff
@@ -133,51 +136,50 @@ def available_designers
 
-  def rescue_has_instances_exception(exception)
-    redirect_back fallback_location: projects_url, alert: "This project has project instances. Delete its project instances first"
+  def rescue_has_instances_exception(_exception)
```
Honestly, I don't know because that isn't my code - I just ran the Rubocop formatter on it! It doesn't look like it needs an argument though, you're right

stanleypliu commented on app/javascript/components/charts/ChartDoughnut.vue:
```diff
@@ -0,0 +1,93 @@
+        options: this.generalOptions
+      })
+    },
+    setStyleOptions() {
```
It is setting style options though because it's setting the background colour and border rather than getting them from somewhere else

WKocur replied:
It returns an object with the styling options, but these options are not actually 'set' until line 81. Or at least that's how I would look at this. 

If you were to call this method in customiseDataset and do nothing else, then the styling options wouldn't have been set. They're only set because of the logic where you spread them into to formattedStats. So I see it as customiseDataset gets the styleOptions and then sets them in formattedStats.

stanleypliu commented on app/javascript/packs/application.js:
```diff
@@ -0,0 +1,25 @@
+import ChartDoughnut from '../components/charts/ChartDoughnut.vue'
+import KpiPage from '../components/pages/KpiPage.vue'
+
+export const eventHub = new Vue()
```
Good spot

stanleypliu commented on lib/modules/kpi/codebase_importer.rb:
```diff
@@ -0,0 +1,56 @@
+
+module Kpi::CodebaseImporter
+  CODEBASE_CREDENTIALS = {
+    username: Rails.application.secrets.codebase_api_username,
+    password: Rails.application.secrets.codebase_api_key
```
I guess we could but it's intended to be your own Codebase username and API key which will have access to the various Codebase hubs of the whole organisation.

WKocur replied:
So what will you use on production and staging?

stanleypliu commented on app/views/kpis/index.html.erb:
```diff
@@ -0,0 +1,5 @@
+<div id="v-app">
+  <kpi-page
+    endpoint="/kpis/kpi_endpoint"
```
Yes, it would be `kpis_kpi_endpoint_path` but that's just as verbose without really helping - also I'm injecting ERB into the prop which I wanted to avoid.

stanleypliu replied:
I've started to become averse to injecting Ruby as props in Vue components as that's what we currently do, but the industry practice seems to be to treat Rails as a backend that we query in methods within the component instead of directly piping Ruby variables into the props and having them available there. I wanted to start practicing decoupling of the FE from the BE with this feature. Also, I've started to try and avoid too much ERB in the view files as it clutters them up. 

You're right about the single source of truth though - if the endpoint was changed for whatever reason then this component will break. I've forgotten that you can alias routes, so provided that alias isn't changed then changing the relative path should be OK. Will fix

stanleypliu commented on lib/modules/kpi/snyk_statistics_importer.rb:
```diff
@@ -0,0 +1,82 @@
+
+      Rails.logger.info("Successfully fetched SVG for #{project.title}")
+
+      svg_text << Nokogiri::HTML(svg).text
```
Slow bit is Snyk having to test the project which it does when we try to access the SVG through HTTParty as it's being dynamically generated which causes a delay (it's not initially accessible), only then when it's finished generating can we start parsing it which is pretty quick.

stanleypliu commented:
> Looks top! Few comments:
> 
> * [x]  I've added a documentation link to one site, but the graph still shows none have documentation (I added this to species+)
> * [x]  Can we use type=number for the number inputs? I don't know if number validation is on the backend, but means we can safely do sums etc. on this data
> * [ ]  Kpi -> KPI
> * [ ]  Can the KPI field be 'Key Performance Indicators (KPIs)'
> * [ ]  Rename KPI measurement to 'KPIs measurement (how and where)'
> * [ ]  Change % project led fields to select with 3 options (Informatics lead, co-design, in-house agency'

About your second point, I've set them to use `number_field` instead of `text_field` which is a crude method of validation (doesn't allow you to add non-whole numbers - will this cause an issue?) 

Re: changing the KPI field in the form to  'Key Performance Indicators (KPIs)', do you want me to make it an array so that you can add more than one KPI per project?

WKocur commented on app/javascript/components/charts/ChartDoughnut.vue:
```diff
@@ -0,0 +1,93 @@
+    customiseDataset() {
+      const correctedLabels = Object.keys(this.statistics).map((key) => {
+        return key.replace(/_/g, ' ').split(' ').map((s) => {
+          return s.charAt(0).toUpperCase() + s.substring(1)
```
Much easier to read - now that I can see what's going on better  I reckon you could skip a loop and do this:

```
      const correctedLabels = Object.keys(this.statistics).map(key => 
        this.capitaliseString(this.replaceUnderscores(key))
      ).join(' ')
```

WKocur replied:
(I haven't tested this though...)

stanleypliu replied:
Just went with lodash's capitalize function - now I've reduced the number of lines of code

WKocur commented:
> > Looks top! Few comments:
> > 
> > * [x]   I've added a documentation link to one site, but the graph still shows none have documentation (I added this to species+)
> > * [x]   Can we use type=number for the number inputs? I don't know if number validation is on the backend, but means we can safely do sums etc. on this data
> > * [ ]   Kpi -> KPI
> > * [ ]   Can the KPI field be 'Key Performance Indicators (KPIs)'
> > * [ ]   Rename KPI measurement to 'KPIs measurement (how and where)'
> > * [ ]   Change % project led fields to select with 3 options (Informatics lead, co-design, in-house agency'
> 
> About your second point, I've set them to use `number_field` instead of `text_field` which is a crude method of validation (doesn't allow you to add non-whole numbers - will this cause an issue?)
> 
> Re: changing the KPI field in the form to 'Key Performance Indicators (KPIs)', do you want me to make it an array so that you can add more than one KPI per project?
> 
> EDIT: On second thought, that might not be so easy to do given that I would have to change the other two associated fields (e.g. adding a checkbox for each KPI which makes it a lot more complex to represent on a chart)

Ah good point on number field, maybe best to leave as text then, unless the answer has to be an integer for some reason. Yearly updates could for example be 0.5 if it were done every 2 years...

For KPIS, I think it's fine for this to just be a text area where you can detail what the KPIs are - keep it flexible and simple I reckon.

WKocur commented on app/models/kpi.rb:
```diff
@@ -82,67 +82,61 @@ def self.api_imports
   end
 
   def self.currently_active_products
     active_projects = Project.where(state: ACTIVE_STATUSES).where.not(last_commit_date: nil).count
```
I think this would be better as an 'or':

Where status is active OR there is a commit in the last year.

We have some active projects for which there is no github repo :(

WKocur commented on app/models/kpi.rb:
```diff
@@ -82,67 +82,61 @@ def self.api_imports
+        hash['0 to 10'] += 1
       else
-        more_than_ten += 1
+        hash['10 to 20'] += 1
```
Might want to change these bins, but fine to leave for now :)

WKocur commented on config/application.rb:
```diff
@@ -40,7 +40,11 @@ class Application < Rails::Application
+    config.autoload_paths += %W(
+      #{config.root}/lib 
+      #{config.root}/lib/modules 
+      #{config.root}/app/serializers
```
I thought serializers were loaded by default - is this not the case?

stanleypliu replied:
Unfortunately not, if they aren't included in the autoloaded paths, then they are not converted to constants and any references to them in other files will result in Uninitialized Constant errors

WKocur replied:
Strange, for me it seems to work fine if I remove the line. We also have serializers in Encore, but are not explicitly autoloading them as far as I can tell. Could it be something to do with restarting the server or perhaps having to restart terminal in order to trigger the loading? I often have to explicitly load a file or reload terminal if something in lib/modules is changed for example.

stanleypliu replied:
Yeah it's weird, it does work without it. I'm going to remove it for now.  The `lib` folders though would need to stay in though as they're not part of `app` so wouldn't be eager or auto loaded.

WKocur commented:
<img width="1340" alt="Screenshot 2020-12-17 at 08 51 35" src="https://user-images.githubusercontent.com/43963744/102465046-1427b600-4045-11eb-84c1-e4283c65894a.png">
Getting this error when updating a project. Seems `Kpi.instance.project_breakdown` is nil.
